pkg/descheduler/types.go

```diff
@@ -0,0 +1,208 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package descheduler
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	fakeclientset "k8s.io/client-go/kubernetes/fake"
+	listersv1 "k8s.io/client-go/listers/core/v1"
+	schedulingv1 "k8s.io/client-go/listers/scheduling/v1"
+	"k8s.io/client-go/tools/events"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/descheduler/cmd/descheduler/app/options"
+	"sigs.k8s.io/descheduler/metrics"
+	"sigs.k8s.io/descheduler/pkg/api"
+	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
+	podutil "sigs.k8s.io/descheduler/pkg/descheduler/pod"
+	"sigs.k8s.io/descheduler/pkg/framework/pluginregistry"
+	frameworkprofile "sigs.k8s.io/descheduler/pkg/framework/profile"
+	frameworktypes "sigs.k8s.io/descheduler/pkg/framework/types"
+	"sigs.k8s.io/descheduler/pkg/utils"
+)
+
+type Descheduler struct {
+	rs                         *options.DeschedulerServer
+	NodeSelector               string
+	PodLister                  listersv1.PodLister
+	NodeLister                 listersv1.NodeLister
+	NamespaceLister            listersv1.NamespaceLister
+	PriorityClassLister        schedulingv1.PriorityClassLister
+	GetPodsAssignedToNode      podutil.GetPodsAssignedToNodeFunc
+	CycleSharedInformerFactory informers.SharedInformerFactory
+	EvictionPolicyGroupVersion string
+	DeschedulerPolicy          *api.DeschedulerPolicy
+	EventRecorder              events.EventRecorder
+}
+
+func NewDescheduler(ctx context.Context, rs *options.DeschedulerServer, deschedulerPolicy *api.DeschedulerPolicy, evictionPolicyGroupVersion string) (*Descheduler, error) {
+	sharedInformerFactory := informers.NewSharedInformerFactory(rs.Client, 0)
+	podInformer := sharedInformerFactory.Core().V1().Pods().Informer()
+	podLister := sharedInformerFactory.Core().V1().Pods().Lister()
+	nodeLister := sharedInformerFactory.Core().V1().Nodes().Lister()
+	namespaceLister := sharedInformerFactory.Core().V1().Namespaces().Lister()
+	priorityClassLister := sharedInformerFactory.Scheduling().V1().PriorityClasses().Lister()
+
+	getPodsAssignedToNode, err := podutil.BuildGetPodsAssignedToNodeFunc(podInformer)
+	if err != nil {
+		return nil, fmt.Errorf("build get pods assigned to node function error: %v", err)
+	}
+
+	sharedInformerFactory.Start(ctx.Done())
+	sharedInformerFactory.WaitForCacheSync(ctx.Done())
+
+	var nodeSelector string
+	if deschedulerPolicy.NodeSelector != nil {
+		nodeSelector = *deschedulerPolicy.NodeSelector
+	}
+
+	var eventClient clientset.Interface
+	if rs.DryRun {
+		eventClient = fakeclientset.NewSimpleClientset()
+	} else {
+		eventClient = rs.Client
+	}
+
+	eventBroadcaster, eventRecorder := utils.GetRecorderAndBroadcaster(ctx, eventClient)
+	defer eventBroadcaster.Shutdown()
```

Comment:
This needs to be invoked outside of `NewDescheduler`

Reply:
Very true! This would have cause another leak